palace/linalg/hypre.cpp

```diff
@@ -47,8 +47,24 @@ HypreCSRMatrix::HypreCSRMatrix(hypre_CSRMatrix *mat) : mat(mat)
   width = hypre_CSRMatrixNumCols(mat);
 }
 
+HypreCSRMatrix::HypreCSRMatrix(mfem::SparseMatrix &m)
+  : palace::Operator(m.Height(), m.Width())
+{
+  mat = hypre_CSRMatrixCreate(height, width, m.NumNonZeroElems());
+  hypre_CSRMatrixI(mat) = m.ReadWriteI();
+  hypre_CSRMatrixJ(mat) = m.ReadWriteJ();
+  hypre_CSRMatrixData(mat) = m.ReadWriteData();
+
+  hypre_CSRMatrixInitialize(mat);
+}
+
 HypreCSRMatrix::~HypreCSRMatrix()
 {
+  if (hypre_CSRMatrixOwnsData(mat) == false)
+  {
+    hypre_CSRMatrixI(mat) = nullptr;
+    hypre_CSRMatrixRownnz(mat) = nullptr;
```

Comment:
I think this still needs to get free'd, as we aren't allocating or freeing it?

Reply:
I see that my comment on the previous never posted for some reason, but here's the rub on using `hypre_CSRMatrixOwnsData(mat)` as the flag for deleting `I` and `NNZ`.

Then that makes no difference from calling `hypre_CSRMatrixDestroy(mat);`. There's 3 scenarios:
1. Built from `*mat` where `hypre_CSRMatrixOwnsData(mat) == true` -> I, J, Data, NNZ all destroyed in `hypre_CSRMatrixDestroy`.
2. Built from `*mat` where `hypre_CSRMatrixOwnsData(mat) ==false` -> Before this, I and NNZ would be deleted by `hypre_CSRMatrixDestroy`, but not J or Data. After this change, they would not be deleted anywhere. 
3. Built from `&m`, then `hypre_CSRMatrixOwnsData(mat) == false`, and I and NNZ are not deleted here, as they're owned by `m` in the first place.

The use of `hypre_CSRMatrixOwnsData(mat)` as the flag introduces the conflict between 2 and 3. If using 
```
      HYPRE_MemoryLocation memory_location = hypre_CSRMatrixMemoryLocation(mat);
      hypre_TFree(hypre_CSRMatrixI(mat),      memory_location);
      hypre_TFree(hypre_CSRMatrixRownnz(mat), memory_location);
 ```
 as in `HYPRE_Int hypre_CSRMatrixDestroy( hypre_CSRMatrix *matrix )`, then the non-owning of `m`, and that it is a `mfem::SparseMatrix` is being ignored. A flag of `bool hypre_managed_memory;` set to false if entering through the `m` channel would suffice, but I had assumed based on your above that you thought that unnecessary, but have gone ahead and put it in.